File: 103.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        res = []
        level = [root]
        flag = 1
        while root and level:
            res.append([node.val for node in level][::flag])
            temp = [(node.left, node.right) for node in level]
            level = [leaf for LR in temp for leaf in LR if leaf]
            flag *= -1
        return res

File: 103.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (root == NULL){
            return {};
        }
        vector<vector<int>> res;
        queue<TreeNode*> q({root});
        int cnt = 0;
        while(!q.empty()){
            vector<int> tmp;
            for (int i = q.size(); i>0; --i){
                TreeNode* node = q.front();
                q.pop();
                tmp.push_back(node->val);
                if (node->left != NULL){
                    q.push(node->left);
                }
                if (node->right != NULL){
                    q.push(node->right);
                }
            }
            if (cnt % 2 == 1){
                reverse(tmp.begin(), tmp.end());
            }
            res.push_back(tmp);
            cnt++;
            
        }
        return res;
    }
};
